# Collatz Sequences

 After seeing a you-tube video about Collatz Sequences, I decided it would be a fun project for coding practice. There are a number of concepts to pull together in this project:
     Python programming- generators, scripting
     General Programming- file directories, environments, packages
     Graph (network) rendering
     Using Jupyter Lab to document my progress

I have written a script in python for calculation of collatz sets; Collatz.py
This script uses the numpy library, so numpy package must be installed in the environment before importing Collatz

In [1]:
import numpy
import Collatz

THere is a known loop in the collatz sequences 4,2,1,4,2,1.... so the function which generates the collatz sequence must check if the result has reached 1. The function also checks the list of known solutions to discover if the current input has alread been solved.

In [2]:
Collatz.colSeq??

Signature: Collatz.colSeq(n, x=3, c=1)
Source:   
def colSeq(n, x=3, c=1):
    '''generates a list of the numbers in the Collatz sequence starting at n'''
    sequence = []
    while n != 1 and (in_true_set(n) == False):
        sequence.append(int(n))
        set_of_true.append(int(n))
        n = col(n, x, c) #finds the next value of n in the collatz sequence
    sequence.append(int(n))
    return sequence
File:      c:\users\kd0nm\anacondaprojects.copy04mar2020\mycodes\training\jupyternotebooks\collatz.py
Type:      function


In [3]:
b = Collatz.colSeq(35)
print(b)

[35, 106, 53, 160, 80, 40, 20, 10, 5, 16, 8, 4, 2, 1]


We can view the list of numbers for which we have already calculated a solution. When we calculate another collatz sequence, the list is updated.

In [4]:
Collatz.set_of_true

[35, 106, 53, 160, 80, 40, 20, 10, 5, 16, 8, 4, 2]

In [5]:
c= Collatz.colSeq(256)
print(c)

[256, 128, 64, 32, 16]


In [6]:
Collatz.set_of_true

[35, 106, 53, 160, 80, 40, 20, 10, 5, 16, 8, 4, 2, 256, 128, 64, 32]

In [9]:
Collatz.set_of_true.sort(reverse=True)
Collatz.set_of_true

[256, 160, 128, 106, 80, 64, 53, 40, 35, 32, 20, 16, 10, 8, 5, 4, 2]

This seems to work well enough, but the output is not very nice to look at. For example, it would appear that 160 is in the same sequence as 256, which it is not. It would be better to see a directed graph of the outputs. So, lets try importing a library for handling directed graphs; altgraph

In [10]:
import altgraph

In [11]:
altgraph?

Type:        module
String form: <module 'altgraph' from 'C:\\Users\\kd0nm\\anaconda3\\envs\\Binnis\\lib\\site-packages\\altgraph\\__init__.py'>
File:        c:\users\kd0nm\anaconda3\envs\binnis\lib\site-packages\altgraph\__init__.py
Docstring:  
altgraph - a python graph library

altgraph is a fork of `graphlib <http://pygraphlib.sourceforge.net>`_ tailored
to use newer Python 2.3+ features, including additional support used by the
py2app suite (modulegraph and macholib, specifically).

altgraph is a python based graph (network) representation and manipulation
package.  It has started out as an extension to the
`graph_lib module
<http://www.ece.arizona.edu/~denny/python_nest/graph_lib_1.0.1.html>`_
written by Nathan Denny it has been significantly optimized and expanded.

The :class:`altgraph.Graph.Graph` class is loosely modeled after the
`LEDA <http://www.algorithmic-solutions.com/enleda.htm>`_
(Library of Efficient Datatypes) representation. The library
includes methods for constru